### ISSUE TITLE
InvalidOperationException "System.Diagnostics.Tracing.EventPipe type == null" after .NET 6.0 migration

### ISSUE THREAD

**Issue by semornas** [microsoft/ApplicationInsights-Profiler-AspNetCore#151]
Hello,

We are facing some issues with the ApplicationInsights Profiler while upgrading our ASP.NET Core 3.1 Apps to .NET 6.0. The apps are  hosted on App Service on Linux Containers and profiling was working fine with .NET Core 3.1.

### Repro:
Now that we have .NET 6.0 instead, the Profiler is throwing a System.InvalidOperationException:
- **Message**: System.Diagnostics.Tracing.EventPipe type == null
- **Exception type**: System.InvalidOperationException
- **Failed method**: Microsoft.ApplicationInsights.Profiler.Core.TraceControls.TraceControl.Initialize
- **Problem Id:**  System.InvalidOperationException at Microsoft.ApplicationInsights.Profiler.Core.TraceControls.TraceControl.Initialize
- **Assembly**: Microsoft.ApplicationInsights.Profiler.Core, Version=2.6.2111.1701, Culture=neutral, -PublicKeyToken=b03f5f7f11d50a3a
- **MessageTemplate**: Unexpected exception in the last circle.
- **SourceContext**: Microsoft.ApplicationInsights.Profiler.Core.Orchestration.MemoryMonitoringSchedulingPolicy
- **Stack trace**:
```
System.InvalidOperationException:
   at Microsoft.ApplicationInsights.Profiler.Core.TraceControls.TraceControl.Initialize (Microsoft.ApplicationInsights.Profiler.Core, Version=2.6.2111.1701, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a)
   at Microsoft.ApplicationInsights.Profiler.Core.TraceControls.TraceControl..ctor (Microsoft.ApplicationInsights.Profiler.Core, Version=2.6.2111.1701, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a)
   at System.RuntimeMethodHandle.InvokeMethod (System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Reflection.RuntimeConstructorInfo.Invoke (System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitConstructor (Microsoft.Extensions.DependencyInjection, Version=6.0.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSiteMain (Microsoft.Extensions.DependencyInjection, Version=6.0.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitRootCache (Microsoft.Extensions.DependencyInjection, Version=6.0.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite (Microsoft.Extensions.DependencyInjection, Version=6.0.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.Resolve (Microsoft.Extensions.DependencyInjection, Version=6.0.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60)
   at Microsoft.Extensions.DependencyInjection.ServiceProvider.CreateServiceAccessor (Microsoft.Extensions.DependencyInjection, Version=6.0.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60)
   at System.Collections.Concurrent.ConcurrentDictionary`2.GetOrAdd (System.Collections.Concurrent, Version=6.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a)
   at Microsoft.Extensions.DependencyInjection.ServiceProvider.GetService (Microsoft.Extensions.DependencyInjection, Version=6.0.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngineScope.GetService (Microsoft.Extensions.DependencyInjection, Version=6.0.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60)
   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService (Microsoft.Extensions.DependencyInjection.Abstractions, Version=6.0.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60)
   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService (Microsoft.Extensions.DependencyInjection.Abstractions, Version=6.0.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60)
   at Microsoft.Extensions.DependencyInjection.ServiceCollectionBuilder+<>c__DisplayClass2_0.<RegisterTraceControls>b__2 (Microsoft.ApplicationInsights.Profiler.AspNetCore, Version=2.6.2111.1701, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a)
   at System.Lazy`1.ViaFactory (System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw (System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Lazy`1.CreateValue (System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at Microsoft.ApplicationInsights.Profiler.Core.TraceControls.AutoFallbackTraceControl.Enable (Microsoft.ApplicationInsights.Profiler.Core, Version=2.6.2111.1701, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a)
   at Microsoft.ApplicationInsights.Profiler.Core.ServiceProfilerProvider+<StartServiceProfilerAsync>d__1.MoveNext (Microsoft.ApplicationInsights.Profiler.Core, Version=2.6.2111.1701, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw (System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at Microsoft.ApplicationInsights.Profiler.Core.Orchestration.OrchestratorEventPipe+<StartProfilingAsync>d__2.MoveNext (Microsoft.ServiceProfiler.EventPipe.Orchestration, Version=2.6.2111.1701, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw (System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at Microsoft.ServiceProfiler.Orchestration.SchedulingPolicy+<StartPolicyAsync>d__1.MoveNext (Microsoft.ServiceProfiler.Orchestration, Version=2.6.2111.1701, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a)
```

Looking at the disassembled code of Microsoft.ApplicationInsights.Profiler.Core I saw the following lines in the TraceControl class where the exception comes from:
```cs
Type type = !(assembly == (Assembly) null) ? assembly.GetType("System.Diagnostics.Tracing.EventPipe") : throw new InvalidOperationException("System.Private.CoreLib assembly == null");
this.m_enableMethod = !(type == (Type) null) ? type.GetMethod("Enable", BindingFlags.Static | BindingFlags.NonPublic) : throw new InvalidOperationException("System.Diagnostics.Tracing.EventPipe type == null");
```
The exception is therefore thrown when `assembly.GetType("System.Diagnostics.Tracing.EventPipe")` returns null, which is the case now in .NET 6.0 (I believe this relates to this PR: https://github.com/dotnet/runtime/pull/48053)

### Package versions:
- Microsoft.ApplicationInsights.AspNetCore: 2.20.0
- Microsoft.ApplicationInsights.Profiler.AspNetCore: 2.3.1-beta3 (similar behavior in 2.3.0)

### Workaround
No workaround identified

Is the use of the Profiler broken in .NET 6.0 or do we need to do something to make it work?

Thanks!

**Comment by xiaomi7732** [comment]
Hi @semornas thanks for your report. Let me take a look.

**Comment by xiaomi7732** [comment]
Hey @semornas I did some investigation. This code path is there for falling back on old .NET Core runtime (like 2.1 or so). It doesn't feel right for an application on .NET 6 to run into it. I test it quickly on my box and I couldn't repro it.

Maybe there's something going on with the runtime image for the container? Do you have a reproduce that I can use for investigation, or could you please let me know which base image did you use that we can try to reproduce it ourselves? Thanks!

**Comment by xiaomi7732** [comment]
Hey @semornas Thanks for sending the info offline. Just to put it on-record, here's some info for debugging:
> the runtime of the apps is Microsoft.AspNetCore.App 6.0.1 (base image: mcr.microsoft.com/dotnet/aspnet:6.0.1-alpine3.14) and the apps are built using  .NET SDK 6.0.101 (base image: mcr.microsoft.com/dotnet/sdk:6.0.101-alpine3.14)
Based on that, I created a repo trying to reproduce the issue - it uses the exact same image while I am not seeing the error.

Could you please help check to see what is that that we are missing there? [Click to open the repo](https://github.com/xiaomi7732/TryReproProfilerIssueOnNet6).

**Comment by semornas** [comment]
Thanks for your help @xiaomi7732; I pushed some additional code in your repo that should expose the issue in local debugging. Hope it will help troubleshooting :)

**Comment by xiaomi7732** [comment]
Thanks for that! Let me take a look.

**Comment by xiaomi7732** [comment]
Thanks for the information. I see what is going on. I have put a workaround in the repo, please [check it out](https://github.com/xiaomi7732/TryReproProfilerIssueOnNet6/commit/8cdbe12faee83e84ef9b5c0f9e116a4f50e4ca82).

Do I understand the intention to pre-warm up singletons is to gain better perf or fail faster?

Here's, I think, what happened:
In Profiler implementation, we have 2 types do the same thing:
**TraceControl** - for backward compatibility works up to .NET 5.0 until this change <https://github.com/dotnet/runtime/pull/48053>.
**DiagnosticsClientTraceControl** - newer implementation relies on newer version of .NET (Core 3.1, 5.x, 6.x ..) runtime.
They both are registered as singleton.

Now, normally, TraceControl, although registered, will only be injected lazily, and dynamically when `DiagnosticsClientTraceControl` fails, that's a strong indication that the app is running on an older version of runtime, and thus the fallback. 

The way the lazy load works rely on that `TraceControl` won't be created when `DiagnosticsClientTraceControl` works.

However, in this case, the warmup service appended to the project creates instances for every singleton ahead of time, and it includes `TraceControl`; Now, `TraceControl` rely on something that is removed from the runtime and thus the exception.

For .NET 6 applications, we don't really need `TraceControl`, so the work around we put there is to skip warm up that one.

We will need think through different scenarios to make a proper fix, and this issue will be kept open until a fix is available.

What do you think @semornas?

**Comment by semornas** [comment]
Hi @xiaomi7732,

Again, thank you for your help! The explanation makes complete sense; this is only because we try to activate all registered singletons (indeed we do that in order to optimize the performances of the application) at startup, that we run into this issue.

The proposed workaround, which is to not activate the `Microsoft.ApplicationInsights.Profiler.Core.TraceControls.TraceControl` type, is perfectly acceptable to mitigate our problem.
Thanks !

**Comment by xiaomi7732** [comment]
Hi @semornas I am glad to hear that. Just so that you know, we landed on a fix to remove the need for the workaround. Please pay attention to the next release.
Again, I am keeping this issue open until the fix is released.

**Comment by xiaomi7732** [comment]
Hi @semornas The fix is released in 2.3.1-beta4 here:
https://www.nuget.org/packages/Microsoft.ApplicationInsights.Profiler.AspNetCore/2.3.1-beta4
